factorio: download using token, not password

elitak opened the pull request:
Downloads were broken by upstream devs' addition of CAPTCHA to the login form.

Changed all references of password to token and used the new GET request + query string to fetch the files.

Fixes #46267

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

elitak committed:
factorio: download using token, not password

Downloads were broken by upstream devs' addition of CAPTCHA to the login
form. We now need only a slightly modified fetchurl to retrieve the
binary distribution.

nbp commented on pkgs/games/factorio/fetch.sh:
```diff
@@ -14,42 +14,32 @@ curl="curl \
+{
+  packageOverrides = pkgs: rec {
+    factorio = pkgs.factorio.override {
+      username = "<username or email address>";
```
nit: Insert a comment saying that this is confidential data and that it should not be shared.
follow-up: Use overlays instead of packageOverrides.

elitak commented:
We can do without the customized `fetchurl` now, but I would like to find a way to inject the error message. The only method I can think of is to nest overrideDerivation alongside override to add a `postBuild` hook, but that gets pretty messy.

elitak commented:
I figured out I should use `preHook` and `failureHook`. I added the info about factorio.com/profile. I will squash these commits after review, but they are close to final, now.

Mic92 commented:
@GrahamcOfBorg eval

elitak committed:
factorio: experimental=true includes stable builds, when newer

elitak commented:
I removed the warning now that 403 is returned, and changed the semantics of the `experimental` package function option to include stable builds.

@GrahamcOfBorg eval

joachifm commented:
Is this ready?

elitak commented:
Yes.
